python_modules/dagster/dagster/_core/execution/asset_backfill.py

```diff
@@ -1711,6 +1711,69 @@ def _should_backfill_atomic_asset_subset_unit(
+
+            if is_self_dependency:
+                self_dependent_node = asset_graph.get(asset_key)
+                # ensure that we don't produce more than max_partitions_per_run partitions
+                # if a backfill policy is set
+                if (
+                    self_dependent_node.backfill_policy is not None
+                    and self_dependent_node.backfill_policy.max_partitions_per_run is not None
+                ):
+                    num_partitions_already_being_requested_this_tick = (
+                        asset_graph_view.get_entity_subset_from_asset_graph_subset(
+                            asset_graph_subset_matched_so_far, asset_key
+                        )
+                    ).size
+
+                    # only the first N partitions can be requested
+                    num_allowed_partitions = max(
+                        0,
+                        self_dependent_node.backfill_policy.max_partitions_per_run
+                        - num_partitions_already_being_requested_this_tick,
+                    )
+                    # TODO add a method for paginating through the keys in order
+                    # and returning the first N instead of listing all of them
+                    # (can't use expensively_compute_asset_partitions because it returns
+                    # an unordered set)
+                    internal_value = entity_subset_to_filter.get_internal_value()
+                    partition_keys_to_include = (
+                        list(internal_value.get_partition_keys())
+                        if isinstance(internal_value, PartitionsSubset)
+                        else [None]
+                    )[:num_allowed_partitions]
+                    partition_subset_to_include = AssetGraphSubset.from_asset_partition_set(
+                        {
+                            AssetKeyPartitionKey(self_dependent_node.key, partition_key)
+                            for partition_key in partition_keys_to_include
+                        },
+                        asset_graph=asset_graph,
+                    )
+                    entity_subset_to_include = (
+                        asset_graph_view.get_entity_subset_from_asset_graph_subset(
+                            partition_subset_to_include, self_dependent_node.key
+                        )
+                    )
+
+                    entity_subset_to_reject = entity_subset_to_filter.compute_difference(
+                        entity_subset_to_include
+                    )
+
+                    if not entity_subset_to_reject.is_empty:
+                        failure_subsets_with_reasons.append(
+                            (
+                                (
+                                    entity_subset_to_filter.compute_difference(
+                                        entity_subset_to_include
+                                    )
+                                ).get_internal_value(),
```

Comment:
```suggestion
                                   entity_subset_to_reject.get_internal_value(),
```
nit - idk if computing these differences is expensive for large subsets

Reply:
once we're working with EntitySubsets differences should generally be cheap - did you have another option in mind?

Reply:
oh wait I'm repeating it, haha, I can definitely fix that